### PR TITLE
Metrics: add ipv4/6 addresses_in_use_total and addresses_total

### PR DESCRIPTION
**Pull request by woodgear**
issue: https://github.com/metallb/metallb/issues/2150

### CODE REVIEW

**Comment by fedepaol** [comment]
Sorry for the delay. I think this makes sense, and would even align with https://github.com/metallb/metallb/blob/main/design/crd-status.md

WDYT @oribon ?

**Comment by oribon** [comment]
sounds good

**Comment by woodgear** [comment]
cool。 i’will start working on this。cr status and metrics are different。 like prometheus  could not monitor via crd status，so i thought we should add it both。in this pr，i will still focus on the metrics part。

**Comment by fedepaol** [comment]
> cool。 i’will start working on this。cr status and metrics are different。 like prometheus could not monitor via crd status，so i thought we should add it both。in this pr，i will still focus on the metrics part。

Yep it's totally fine to work only on the metrics part.

**Comment by fedepaol** [comment]
Hey @woodgear , just pinging to check if you are still interested to work on this

**Comment by woodgear** [comment]
yes. im try to fighting the ci..

**Comment by woodgear** [comment]
@fedepaol

File: internal/allocator/allocator.go [hunk 1]
```diff
@@ -391,14 +430,17 @@ func sharingOK(existing, new *key) error {
 }
 
 // poolCount returns the number of addresses in the pool.
-func poolCount(p *config.Pool) int64 {
+func poolCount(p *config.Pool) (int64, int64, int64) {
```
Comment (fedepaol) [inline internal/allocator/allocator.go:433]: Can you adjust the godoc?

File: internal/k8s/controllers/reconciliation_test.go [hunk 2]
```diff
@@ -179,7 +179,7 @@ var _ = AfterSuite(func() {
 	cancel()
 	err := testEnv.Stop()
 	Expect(err).NotTo(HaveOccurred())
-	Eventually(mgrDone.Load(), 5*time.Second, 200*time.Millisecond).Should(BeTrue())
+	Eventually(mgrDone.Load, 5*time.Second, 200*time.Millisecond).Should(BeTrue())
```
Comment (fedepaol) [inline internal/k8s/controllers/reconciliation_test.go:182]: Why this change?

File: e2etest/l2tests/l2.go [hunk 3]
```diff
@@ -502,7 +502,7 @@ var _ = ginkgo.Describe("L2", func() {
 				L2Advs: []metallbv1beta1.L2Advertisement{emptyL2Advertisement},
 			}
 
-			poolCount, err := config.PoolCount(resources.Pools[0])
+			total, _, _, err := config.PoolCount(resources.Pools[0])
```
Comment (fedepaol) [inline e2etest/l2tests/l2.go:505]: Can we check v4 and v6 numbers too depending on ipFamily?

**Comment by fedepaol** [comment]
Thanks, a couple of nits but looks really neat!

File: internal/k8s/controllers/reconciliation_test.go (hunk 2 above)
Comment (woodgear) [reply internal/k8s/controllers/reconciliation_test.go:182]: golangci-lint complains about
`
internal/k8s/controllers/reconciliation_test.go:182:2: ginkgo-linter: use a function call in Eventually. This actually checks nothing, because Eventually receives the function returned value, instead of function itself, and this value is never changed; consider using `Eventually(mgrDone.Load, 5*time.Second, 200*time.Millisecond).Should(BeTrue())` instead (ginkgolinter)
`

File: e2etest/l2tests/l2.go (hunk 3 above)
Comment (woodgear) [reply e2etest/l2tests/l2.go:505]: ok

File: internal/allocator/allocator.go (hunk 1 above)
Comment (woodgear) [reply internal/allocator/allocator.go:433]: ok

File: e2etest/l2tests/l2.go [hunk 4]
```diff
@@ -514,21 +514,39 @@ var _ = ginkgo.Describe("L2", func() {
...
+					expect int64
 				}
-				err = metrics.ValidateGaugeValue(int(poolCount), "metallb_allocator_addresses_total", map[string]string{"pool": poolName}, controllerMetrics)
-				if err != nil {
-					return err
-				}
-				err = metrics.ValidateOnPrometheus(promPod, fmt.Sprintf(`metallb_allocator_addresses_in_use_total{pool="%s"} == 0`, poolName), metrics.There)
-				if err != nil {
-					return err
+				cases := Cases{
```
Comment (fedepaol) [inline e2etest/l2tests/l2.go:521]: nit: given it's local, you can use an anonymous struct here (and no need to have it exported anyway).

Something like cases := []struct{name,
                                                          expect}{
                                            }

File: internal/k8s/controllers/reconciliation_test.go (hunk 2 above)
Comment (fedepaol) [reply internal/k8s/controllers/reconciliation_test.go:182]: Not sure you saw my previous comment, but can you move this to a separate commit (still in this pr)?

File: e2etest/l2tests/l2.go (hunk 4 above)
Comment (woodgear) [reply e2etest/l2tests/l2.go]: nice trick.

File: internal/k8s/controllers/reconciliation_test.go (hunk 2 above)
Comment (woodgear) [reply internal/k8s/controllers/reconciliation_test.go:182]: ok

**Comment by woodgear** [comment]
@fedepaol what should i do next to make it merge?

**Comment by fedepaol** [comment]
LGTM, one final ask: https://github.com/metallb/metallb/pull/2151#discussion_r1450404190
Can you move that change in a separate commit (still in this PR)? Thanks a lot!

**Comment by woodgear** [comment]
@fedepaol updated.

**Comment by fedepaol** [comment]
Awesome, sending to the merge queue!
